### DISCUSSION TITLE
Bump the Java version to 11 in SLF4J version 2.1.0

### DISCUSSION THREAD

**Discussion by ceki** [qos-ch/slf4j#379] [Polls]
Moving the Java requirement from Java 8 to Java 11 in SLF4J version 2.1.0 is being considered. 

We are seeking input regarding this change in requirements.

The desire upgrade to Java 11 stems from the difficulty in maintaining both Java 8 and JPMS support in the SLF4J build. The build is rather complicated.

**Update**: Given the widespread use of SLF4J, it stands to reason that SLF4J should stick to Java 8 for the moment. As such, **SLF4J version 2.1.0-alpha1 reverted to using Java 8**.

**Comment by KengoTODA** [comment]
Since Java 11 has reached its end of life, I believe using Java 17 would be a good option too.

**Reply by ceki** [reply]
The upgrade to Java 11 stems from the difficulty in maintaining both Java 8 and JPMS support in the build. It's too much of a headache.

**Reply by hazendaz** [reply]
For reference to jdk 11 end of support -> https://endoflife.date/oracle-jdk

So technically premier support is now ended but Extended is still alive.  However, I doubt that even matters to anyone really except very niche audience that refuses to upgrade.

I'd also say move to 17.  Reason, Spring and spring boot forced that already a year ago.  Another reason, Jakarta is forcing jdk 21 for upcoming release.  The community is really all over the map and I think at least with jdk 17 the newer features that can be used would be highly worth it.  You could always just cut side branches for back support of critical vulnerabilities only for older jdks but I'd say those customers should be paying some extended support ;)

**Reply by ceki** [reply]
@hazendaz Thank you for your input.

Come to think of it, I don't see how upgrading to JDK 17 would be of help in designing a better API.  Do you see how text blocks, switch expressions or records all  new in Java 17 would make a difference in SLF4J API as seen by the end user?

**Reply by hazendaz** [reply]
Probably no benefit to end user.

**Reply by anbusampath** [reply]
As @hazendaz mentioned Spring Boot 3+ require Java 17, also Microunat 4, Quarkus 3.7+ required Java 17. Library authors can stay on Java 11, but if all Java ecosystem libraries and frameworks move to Java 17, it would great from end user perspective.

**Reply by hkupty** [reply]
I think moving to jdk 17 should be prioritised over jdk11, though I see the logic in bumping to 11. One could increase the requirement to jdk 11 in 2.1.y and to 17 in a subsequent 3.x.y series

**Reply by albertus82** [reply]
Logging libraries are probably the most used dependencies of all. IMHO any unnecessary upgrade will limit the adoption of SLF4J and perhaps even Logback.

**Reply by hazendaz** [reply]
If this helps at all, the only other real player in this logging 'wars' as they use to call it is Log4j already made their call, see https://logging.apache.org/log4j/3.x/.

**Reply by xzel23** [reply]
For me, Java 8 is a real pain. Many dependencies I use don't have proper JPMS support, and patches or PRs are sometimes refused because either a JDK 9+ is required for building, or fear that a module-info breaks things for people still on 8. I really welcome each and every library that does the step to at least Java 11. In the case of SLF4J, we already have proper module support (thank you, @ceki), and I don't know if SLF4J users would benefit from a better API that might be possible using 11 or 17. But going to at least 11 would put more pressure on other projects to finally move on.
So I could understand if SLF4J stayed on 8, but I hope for 11 or even better 17.

**Reply by ppkarwasz** [reply]
> If this helps at all, the only other real player in this logging 'wars' as they use to call it is Log4j already made their call, see https://logging.apache.org/log4j/3.x/.

Note that the Java 17 baseline only applies to Log4j Core (the implementation). Log4j API will still be compatible with Java 8 (and there will be no version 3.0.0 for the API to stress out the lack of breaking changes).

**Comment by bmarwell** [comment]
Yes, please do an 11/17 upgrade (11 is almost EOL even with extended support), but please bump to a new major version!

Can't wait for new APIs to be added! 👍🏻

**Reply by bmarwell** [reply]
Quick addition:
I am happy that @ceki stayed on Java 8 for the moment for several reasons.

1. My solution would have meant: Maintaining a 3.x version for Java 11 or 17, as well as still backporting security fixes to Java 8 (v2.x). That is a LOT of work.
2. A logging API (not the implementation!) does not  really benefit a lot from new APIs. I do like the Java (JUL) logging API with `Supplier<String>` -- but honestly? The current template syntax `{}` with a simple logging guard statement does effectively the same. There is no technical reason, a cosmetical reason at most.

Thus, libraries like this are the LAST to be updated to require a new Java version. Even after Libraries like Guava or the Build Tools, like Maven 4 requiring Java 17.

So for the time being, unless THESE numbers drop (see: https://github.com/qos-ch/slf4j/discussions/379#discussioncomment-8120363), Let us stick with 2.x == Java 8. A 3.x with Java 11/17 (probably 17 by then) can follow any time, no?

**Comment by marcwrobel** [comment]
I don't agree with https://github.com/qos-ch/slf4j/discussions/379#discussioncomment-7885576:

> Since Java 11 has reached its end of life, I believe using Java 17 would be a good option too.

and https://github.com/qos-ch/slf4j/discussions/379#discussioncomment-7889972 :

> For reference to jdk 11 end of support -> https://endoflife.date/oracle-jdk

Only mentioning Oracle JDK, which is a commercial Java distribution, is not very representative IMO. Here are a few other dates for some OpenJDK distributions (source : pages on https://endoflife.date/tags/java-distribution):

- Eclipse Temurin : At least Oct. 2027 for both OpenJDK 11 and 17,
- Amazon Corretto: Sept. 2027 for 11, Oct. 2029 for 17,
- Azul Zulu: Sep. 2026 for 11 (Sep. 2028 with commercial support), Sep. 2029 for 17 (Sep. 2031 with commercial support).

Given those dates, I think an upgrade to JDK 11 is still a thing to consider.

I agree that there is no point keeping JDK 8/11 compatibility if almost all major frameworks / libraries raised there compatibility level to 17. But SLF4J being a basic bloc for many applications, I think it should at least lag behind all those frameworks in term of Java compatibility.

So my opinion is: keep the JDK 8 compatibility for now if it's manageable (meaning if it does not become a security risk / if it does not take too much time / or even if it does not prevent doing some nice improvements to the API), or upgrade to 11 otherwise.

**Comment by giacgbj** [comment]
Baseline Java 17 and a critical vulnerability (CVSS>9, at least a remote code execution) in the previous version would be great to force applications' upgrades 😉

**Reply by albertus82** [reply]
...or to migrate to another logging library.

**Reply by xzel23** [reply]
I estimate the switch to another logging library would be more effort than moving to 8, and given that log4j will also move to 17, there's not much to chose from anyway.

**Comment by athspk** [comment]
Anything >= 17

**Comment by henjovr** [comment]
Java 8 is still widely used in big (slow!) enterprises, so unless you want people to migrate away from SLF4j, keep supporting Java 8.
It is an interface and a facade. Keep it simple. In my opinion a logging framework shouldn't need the latest Java features to function correctly. A logger binding - perhaps - but not the facade.

In my opinion, SLF4J is clean, stable and robust as it is now (2.0). Please keep it clean, stable and robust.

Maybe an idea: keep 2.x for Java 8 (and support it), make this breaking (!!) change into a 3.x version.

**Reply by sysmat** [reply]
In my opinion this is valid point, if interface don't use language features from java X why bump it

**Reply by ceki** [reply]
@henjovr Thank you for your comments which make a lot of sense. SLF4J version 2.1.0-alpha1 reverts to Java 8 as a requirement.

**Reply by henjovr** [reply]
Thank for listening, much appreciated :)

**Comment by michael-o** [comment]
SLF4J is a fundemental library, therefore it should be fundamentally available to everyone and not to fraction of people: Java 8+ for everyone.

**Reply by bmarwell** [reply]
-1. Fundamental does not mean "supports unsupported Java versions". I mean, there's still 1.x and 2.x if Ceki accepts to do a 3.x version for the Java 17 update. See: https://github.com/qos-ch/slf4j/discussions/379#discussioncomment-8080681

For me, this (available to everyone) is neither an argument nor a technical justification. I don't have (much) pity for those still running Java 8 in two years. They will face other problems anyway as other fundamental libraries will move to Java 17, too. slf4j is just one piece of many when maintaining software based on outdated Java versions.

**Reply by michael-o** [reply]
Again, we are *not* talking about not supported Java versions. Java 8 will live until 2030 for many vendors. Stop spreading FUD, whether someone is willing to support Java 8 in his codebase is a different discussion.

**Reply by bmarwell** [reply]
I know that (some vendors do SECURITY updates until 2030, no new features like TLS versions etc). I just wanted to imply that a 3.x version would be beneficial for Java 17+ users as long as v2.x is still being maintained (eg security fixes). i don't see where this is FUD.

**Reply by bmarwell** [reply]
For Oracle Java, even Premier support already ended in 2022: https://www.oracle.com/java/technologies/java-se-support-roadmap.html. They are on extended support, which (don't know, but most probably) means that no new features will be added, like new TLS versions, ciphers, etc.

For IBM Java, it is 2025, which is way sooner: https://www.ibm.com/support/pages/java-sdk-lifecycle-dates

Eclipse Temurin EOS is April 2026 (Azul): https://www.azul.com/products/azul-support-roadmap/. However, they do offer support until 2030, too. 

Based on that, I do not consider my comment to be FUD. However, I am sorry if the wording caused any confusion. CC @albertus82.

**Reply by albertus82** [reply]
If none of the features of JDK 17 are really useful for the development of SLF4J, I think there's no need to start another branch (3.x) that in turn would also introduce a maintenance cost without any added value.

**Reply by bmarwell** [reply]
That's a very fair point.

**Reply by juliojgd** [reply]
> therefore it should be fundamentally available to everyone and not to fraction of people: Java 8+ for everyone.

I would advocate for upgrading at least to Java 11 but this statement is incorrect: "available to everyone and not to fraction of people: Java 8+ for everyone."

Everyone is not using Java 8+. Believe it or not, lots of companies are still using Java < 8. The statement pretends to be all-inclusive but is not, so I think the statement is invalid from the foundations.

If SLF4J intent would be to include "everyone" it should support any version of Java used for at least one  developer, let it be Java 1.1, 1.2, 1.7, etc. It seems that the original idea is to be available to a part of Java versions only. No change in the directive if the upgrade to Java 11 (or 17) is made.

**Comment by ijuma** [comment]
Another point of view is that there is slf4j 1.x for people who are on older Java versions. Generally, the ecosystem is moving on, see https://youtu.be/NxpHg_GzpnY?si=wA57g9kAhYulrlUO&t=411

**Reply by michael-o** [reply]
Then the lib should move to an honest 3.x.

**Comment by opwvhk** [comment]
As mentioned by others, Java 8 won't be gone just yet. In fact, Azul Zulu 8 will be supported until the end of **2030** (almost 7 more years!), with security fixes for 2 more years after that. So come up with actual usage statistics or expect people to migrate to something else.

However, supporting Java 8 does not mean that there is no room for innovation! One option is to use a [Multi-Release JAR (JEP 238)](https://openjdk.org/jeps/238).

The advantage of that is that SLF4J will still work with Java 8, but there is the option of more functionality for Java 9, 11, 17, ..., such as an implementation of `System.Logger`.

**Reply by froque** [reply]
There are some statistics by Jetbrains: https://www.jetbrains.com/lp/devecosystem-2023/java/

**Reply by bmarwell** [reply]
Oh. That's a very sad result.

**Reply by michael-o** [reply]
That is reality, not Willy Wonka's factory.

**Reply by sysmat** [reply]
@opwvhk yes Multi-Release JAR Files I think this is very good

**Reply by Geolykt** [reply]
Just for your information, SLF4J does already use multi-release jars for the module-infos. From my experience doing so isn't fun though as buildsystems are really bad at combining Java 8, JPMS (also known as Jigsaw) and MRJ. Ultimately this is a write-once-forget-forever kind of situation. Especially maven does treat that system a little bit better than Gradle but you need to do a non-zero amount of setup regardless.

**Reply by hkupty** [reply]
Just a note on Java8 usage statistics:
[2019](https://www.jetbrains.com/lp/devecosystem-2019/java/) -> 83%
[2020](https://www.jetbrains.com/lp/devecosystem-2020/java/) -> 75%
[2021](https://www.jetbrains.com/lp/devecosystem-2021/java/) -> 72%
[2022](https://www.jetbrains.com/lp/devecosystem-2022/java/) -> 60%
[2023](https://www.jetbrains.com/lp/devecosystem-2023/java/) -> 50%

It is still relevant, but losing relevance rapidly...

**Reply by xzel23** [reply]
... and I bet that 90% of the developers still using Java 8 on a regular basis hope that some important library they need switches to at least 11 so that they can convince their management to update ;-)

**Reply by albertus82** [reply]
Nope, the management will convince 'em to switch to another library.

**Reply by xzel23** [reply]
Ah, that one should be easy. The other big logging library has already announced they will switch to 17. They can tell management there are no widely used logging libraries left that still support Java 8. Even JUL logging is on Java 21 with the latest release (just don't tell management that it's part of the JDK and so JDK 8 support will be there as long as Java 8 is around).

**Comment by hkupty** [comment]
In the context of a logging façade library, bumping the java version means that java 8 apps will be forbidden to use the newer APIs, period. Writing logs is the responsibility of the underlying libraries, like logback, log4j or [penna](https://github.com/hkupty/penna) (this last one already caps the support to java 17 anyway).

It doesn't pose any immediate requirements to apps running java 8 to upgrade and the short-term impact is that they will not be able to refactor to use the newer API. Longer term, they will not be able to update dependencies that use newer SLF4J version.

I wonder how many apps running java 8 are keeping all the dependencies up-to-date to really feel the impact.

The concerns feel much worse in theory than in practice it seems to me..

**Reply by Geolykt** [reply]
As someone that actively develops a program (well it's actually a modification to one, but for the sake of this argument that is not all that relevant) for Java 8 I can actually say that I do intend on keeping all my libraries up-to-date even though I use Java 8.

Yes, SLF4J not supporting Java 8 in the future is a minor issue for all API consumers, but as others raised doing so in version 2.1.0 is inappropriate. Outside of that it is my firm belief that all API providers should support as many systems as possible, while still remaining within the bounds of sanity.  One could make the argument that you can "downgrade" Java 9+ artifacts to Java 8 - projects for that exist. Whether that should happen when compiling SLF4J or when compiling whatever project consumes it is up to the maintainers - both ways have their drawbacks (and they aren't perfect anyways - pick your poison).

**Reply by hkupty** [reply]
> but as others raised doing so in version 2.1.0 is inappropriate

I can agree with that.

> all API providers should support as many systems as possible

I see your point, but I can think of cases where that would be impractical. I don't think that's the case for SLF4J though.

**Comment by msgilligan** [comment]
@ceki I'm glad you're staying with Java 8. I normally lean towards trying to make things as modern as possible. But for something as widely-used as logging, supporting a wide range of JDK versions is a good thing. If I'm building an app that uses JDK 21, but one of my dependencies is a JAR that supports JDK 8 and uses SLF4J logging, I'd rather it use the latest version of SLF4J so I don't get a version conflict when I link it in.

> The desire upgrade to Java 11 stems from the difficulty in maintaining both Java 8 and JPMS support in the SLF4J build. The build is rather complicated.

I haven't looked at the build, but maybe you can simplify it by not making it an MR-JAR. There are tools that will just put the `module-info.class` in a JDK 8 JAR. This is done by some very popular/standard JARs without any issues, such as `jakarta.inject-api-2.0.1.jar` and `jakarta.annotation-api-2.0.0.jar`. At one time there were build tools that might have a problem with this, but I think all major tools have been updated.

If you really want to have a `module-info.jar` without being "transgressive" and putting it in a JDK 8 JAR or using an MR JAR tool, you could just bump the version to Java 9. I'm not sure how many libraries there are that support Java 9, but not 11, but in Android apps for example it might be nice to stick with 9, rather than 11.

**Comment by DanskerDave** [comment]
I'd go for Java 17: it's a damn good all-round LTS release.
Java 11 (albeit LTS) is no-mans-land.
Java 8 has had its day (_"and half of the day after"_, as Clive James said).
Maven decided to base Maven 4 on Java 17 a year ago.

**Reply by athspk** [reply]
I would start planning for 17 too instead of 11. 
Java 8 is oddly still around but will soon leave this world 😄 
According to [this report](https://www.scribd.com/document/730660713/Report-Jrebel-2024-Dev-Productivity), Java 8 had 24% usage last year.

**Reply by michael-o** [reply]
Yet another report which doesn't represent reality and isn't freely available...

**Reply by Geolykt** [reply]
Either way, 24% usage is still quite the large number

**Reply by DanskerDave** [reply]
24% is shockingly large when you consider the thing is 11 years old.
Worldwide, people are being forced to develop with this museum piece.
There's so much good stuff in Java 17.
Java 11, by comparison, is half finished.

**Reply by michael-o** [reply]
> 24% is shockingly large when you consider the thing is 11 years old. Worldwide, people are being forced to develop with this museum piece. There's so much good stuff in Java 17. Java 11, by comparison, is half finished.

What is the problem?  Larger setups develop software and have it run 10+ years.  What is the purpose everything constantly?

**Reply by ppkarwasz** [reply]
> Worldwide, people are being forced to develop with this museum piece. There's so much good stuff in Java 17.

Is there any new feature in Java 17 that could be used in SLF4J? I don't think so, `java.util.function.Supplier` was the last enhancement that could be applied to a logging API.

[`ScopedValue`](https://docs.oracle.com/en/java/javase/24/docs/api/java.base/java/lang/ScopedValue.html) might be the next one and a `ScopedValue`-compatible interface could replace the aging `MDC` interface. Such an interface, however, should probably be published as a separate artifact from `slf4j-api`.

**Reply by hkupty** [reply]
I think there's a piece of information we're not seeing when we look at those numbers like `24% are still using JDK 8`: How much of that is actually something _new_ being developed? I doubt it's the totality of those 24%. Most if not all of that is maintenance, of which most if not all will keep using old libraries.

A much better heuristic would be to figure out how many applications are using SLF4J 2.x _and_ JDK 8. Those would be the ones that would matter for this decision. If this number is <1%, then there's nothing stopping increasing the language requirement, and, as a practical benefit, allowing the usage of more modern language features and, most importantly, signaling that the JVM ecosystem is moving past JDK 8 finally.

**Comment by albertus82** [comment]
As mentioned, Java 8 will still be supported for years. I say don't upgrade SLF4J to a newer JDK without a real need. If @ceki says 11 is enough, then IMO there's no need to cause more problems for people than necessary.

**Reply by xzel23** [reply]
I don't think the fact that some vendors still offering support for Java 8 should have too much impact on this decision. Whoever choses for whatever reason to stay on an outdated Java release cannot expect to be able to use the latest version of every library. But the main problem IMHO has been solved in the meantime, i.e., SLF4J having full JPMS support with a proper module-info.

As much as I'd personally like to see SLF4J to move beyond Java 8, simply because it will make others who still refuse to implement a proper module-info move as well ("we will support Java 8 as long as possible, and adding module-info might break things or cause work for us"), until there's actually some Java 8 incompatible change to the API, and there are no other reasons, I think we should just be patient.

Maybe it would be a good idea to just follow Eclipse Temurin EOL dates (November 2026 for Java 8, October 2027 for 11 and 17) and communicate those dates early and prominently on the project page so that everybody is aware of it and has enough time to prepare?

**Reply by albertus82** [reply]
I'm a strong supporter of upgrading to the latest LTS version available, but you know the world outside is not the earthly paradise, for example there are many applications that still use log4j v1 because the upgrade is not worth the effort, because maybe someone built an entire logging framework on top of log4j v1, etc. There may be many complex scenarios that we can't even imagine, so IMO the less unnecessary problems you create, the better.

**Comment by victorherraiz** [comment]
SLF4J is so relevant that It could push the industry to move forward. As Spring did, moving the baseline to 17 and forcing outdated applications to update, I think you should do too. Providing a reasonable schedule, of course.

**Reply by Geolykt** [reply]
And what's the point of forcing the industry to push forward? At that point you might also request Java runtime vendors to remove the Java 8 runtimes. Or one might even propose a law for that.

Use whatever you want at runtime, but please don't mess with ABI (Application Binary Interface) without just cause (e.g. a just cause would be "hey, we use JNI but JNI is probably gone in 20 to 30 years so we'll replace it with FFI that hopefully lasts longer than JNI"). If you want to mess with ABI, then batch these changes together and (this is the most important part) change the major version number.

Also what even is a reasonable schedule? And what are you even going to do in that Schedule? Personally, I don't see any point in giving much warning – sure, you want to include that kind of stuff in the changelogs – but otherwise it's still pretty much a 0 or 1 kind of affair: Either it runs on your version of Java or it does not. You could create warning messages but those are easily ignored.

**Reply by victorherraiz** [reply]
Support libraries is a big effort, and it is even bigger when you have to support old versions and, maybe, introduce new fancy stuff now and then. New versions allow new solution that are faster and easier to maintain. If the library fails to do so, a new alternative, with less experience but also with less legacy, it will be an option. 

And this is a nice support schedule https://spring.io/projects/spring-boot#support

**Reply by michael-o** [reply]
@victorherraiz This is an unfair comparison. They make a living out of commercial support. SLF4J is supported for years...

**Comment by gaeljw** [comment]
I'm in the team that votes for bumping to at least 11 if not 17. Even more if this can simplify the build of SLF4J.

Most major libraries/frameworks have moved already to 17.

Users that are still on Java 8 are unlikely to upgrade SLF4J anyway. They can keep using the version they are using today (2.0.x if they are but I'd bet they are even a few versions behind). 

My only minor concern is if this should be considered a breaking change (hence version 3.x) or not. I could live with both.

**Comment by colorado1028mc-hash** [comment]
> Moving the Java requirement from Java 8 to Java 11 in SLF4J version 2.1.0 is being considered.
> 
> We are seeking input regarding this change in requirements.
> 
> The desire upgrade to Java 11 stems from the difficulty in maintaining both Java 8 and JPMS support in the SLF4J build. The build is rather complicated.
> 
> **Update**: Given the widespread use of SLF4J, it stands to reason that SLF4J should stick to Java 8 for the moment. As such, **SLF4J version 2.1.0-alpha1 reverted to using Java 8**.
